pkg/ride/compiler/compaction.go

```diff
@@ -193,3 +193,103 @@ func (c *Compaction) processExpr(node ast.Node) ast.Node {
 		return node
 	}
 }
+
+func getUsedNames(node ast.Node, usedNames map[string]struct{}) {
+	switch expr := node.(type) {
+	case *ast.AssignmentNode:
+		getUsedNames(expr.Expression, usedNames)
+		getUsedNames(expr.Block, usedNames)
+	case *ast.FunctionDeclarationNode:
+		getUsedNames(expr.Body, usedNames)
+		getUsedNames(expr.Block, usedNames)
+	case *ast.FunctionCallNode:
+		usedNames[expr.Function.Name()] = struct{}{}
+		for _, a := range expr.Arguments {
+			getUsedNames(a, usedNames)
+		}
+	case *ast.PropertyNode:
+		getUsedNames(expr.Object, usedNames)
+	case *ast.ConditionalNode:
+		getUsedNames(expr.Condition, usedNames)
+		getUsedNames(expr.TrueExpression, usedNames)
+		getUsedNames(expr.FalseExpression, usedNames)
+	case *ast.ReferenceNode:
+		usedNames[expr.Name] = struct{}{}
+	}
+}
+
+func diff(map1, map2 map[string]struct{}) map[string]struct{} {
+	difference := make(map[string]struct{})
+
+	for key, value := range map1 {
+		if _, exists := map2[key]; !exists {
+			difference[key] = value
+		}
+	}
+
+	return difference
+}
+
+func appendMapNames(to, from map[string]struct{}) map[string]struct{} {
+	res := make(map[string]struct{})
```

Comment:
We can improve by creating new map with a size of `to` and `from` maps combined.
Also, there's a function experimental standard function https://pkg.go.dev/golang.org/x/exp/maps#Copy that makes almost what you need. But it copies keys and values from source to destination maps and doesn't return any map. May be not so useful in this case.